File: skills/architecture-compliance/utils/resolve-includes.py
#!/usr/bin/env python3

"""
Template Include Resolution Utility

Resolves @include and @include-with-config directives in compliance templates.

Usage:
    python3 resolve-includes.py <template-file> [output-file]

If output-file is omitted, outputs to stdout.
"""

import json
import os
import re
import sys
from pathlib import Path

SKILL_DIR = Path(__file__).parent.parent
MAX_DEPTH = 3


def load_config(config_name):
    """Load a JSON config file"""
    config_path = SKILL_DIR / 'shared' / 'config' / f'{config_name}.json'

    if not config_path.exists():
        raise FileNotFoundError(f'Config file not found: {config_path}')

    with open(config_path, 'r', encoding='utf-8') as f:
        return json.load(f)


def replace_variables(content, config):
    """Replace {{variables}} in content with values from config"""

    def replacer(match):
        var_name = match.group(1)
        if var_name in config:
            return str(config[var_name])
        # Keep the placeholder if variable not found in config
        print(f'Warning: Variable {{{{{var_name}}}}} not found in config', file=sys.stderr)
        return match.group(0)

    return re.sub(r'\{\{(\w+)\}\}', replacer, content)


def parse_directive(directive_text):
    """Parse an include directive"""
    # Pattern: <!-- @include(-with-config)? path/to/file.md (config=name)? -->
    pattern = r'<!--\s*@include(-with-config)?\s+(.+?)\s*(?:config=(\S+))?\s*-->'
    match = re.match(pattern, directive_text)

    if not match:
        return None

    return {
        'full': match.group(0),
        'type': 'with-config' if match.group(1) else 'simple',
        'file_path': match.group(2).strip(),
        'config_name': match.group(3) or None
    }


def resolve_include(directive, config, depth, processed_files):
    """Resolve a single include directive"""
    include_type = directive['type']
    file_path = directive['file_path']
    config_name = directive['config_name']

    # Prevent infinite recursion
    if depth > MAX_DEPTH:
        raise RecursionError(f'Max include depth ({MAX_DEPTH}) exceeded')

    # Resolve file path relative to skill directory
    full_path = SKILL_DIR / file_path

    # Detect circular includes
    full_path_str = str(full_path)
    if full_path_str in processed_files:
        raise ValueError(f'Circular include detected: {file_path}')

    if not full_path.exists():
        raise FileNotFoundError(f'Include file not found: {full_path}')

    # Read the included file
    with open(full_path, 'r', encoding='utf-8') as f:
        content = f.read()

    # Mark as processed
    processed_files.add(full_path_str)

    # If it's a parameterized include, load config and replace variables
    if include_type == 'with-config':
        if not config_name:
            raise ValueError(f'Config name required for @include-with-config: {file_path}')

        domain_config = config or load_config(config_name)
        content = replace_variables(content, domain_config)

    # Recursively resolve nested includes
    content = resolve_includes(content, config, depth + 1, processed_files)

    # Remove from processed to allow reuse in different branches
    processed_files.discard(full_path_str)

    return content


def resolve_includes(content, config=None, depth=0, processed_files=None):
    """Resolve all includes in content"""
    if processed_files is None:
        processed_files = set()

    # Find all include directives
    include_pattern = r'<!--\s*@include(-with-config)?\s+.+?\s*(?:config=\S+)?\s*-->'

    result = content
    for match in re.finditer(include_pattern, content):
        directive_text = match.group(0)
        directive = parse_directive(directive_text)

        if not directive:
            print(f'Warning: Could not parse directive: {directive_text}', file=sys.stderr)
            continue

        try:
            resolved_content = resolve_include(directive, config, depth, processed_files)
            result = result.replace(directive['full'], resolved_content, 1)
        except Exception as e:
            print(f"Error resolving include: {directive['file_path']}", file=sys.stderr)
            print(f"  {str(e)}", file=sys.stderr)
            # Keep the directive in place on error

    return result


def main():
    """Main function"""
    if len(sys.argv) < 2:
        print('Usage: python3 resolve-includes.py <template-file> [output-file]', file=sys.stderr)
        print('', file=sys.stderr)
        print('Example:', file=sys.stderr)
        print('  python3 resolve-includes.py templates/TEMPLATE_BUSINESS_CONTINUITY.md expanded.md', file=sys.stderr)
        sys.exit(1)

    template_path = sys.argv[1]
    output_path = sys.argv[2] if len(sys.argv) > 2 else None

    if not os.path.exists(template_path):
        print(f'Error: Template file not found: {template_path}', file=sys.stderr)
        sys.exit(1)

    try:
        # Read template
        with open(template_path, 'r', encoding='utf-8') as f:
            content = f.read()

        # Resolve includes
        expanded = resolve_includes(content)

        # Output
        if output_path:
            with open(output_path, 'w', encoding='utf-8') as f:
                f.write(expanded)

            print(f'Template expanded successfully: {output_path}')

            # Show statistics
            original_lines = len(content.splitlines())
            expanded_lines = len(expanded.splitlines())
            print(f'Lines: {original_lines} → {expanded_lines} (+{expanded_lines - original_lines})')
        else:
            print(expanded)

    except Exception as e:
        print(f'Error: {str(e)}', file=sys.stderr)
        sys.exit(1)


if __name__ == '__main__':
    main()

File: skills/architecture-compliance/utils/resolve-includes.js
#!/usr/bin/env node

/**
 * Template Include Resolution Utility
 *
 * Resolves @include and @include-with-config directives in compliance templates.
 *
 * Usage:
 *   node resolve-includes.js <template-file> [output-file]
 *
 * If output-file is omitted, outputs to stdout.
 */

const fs = require('fs');
const path = require('path');

const SKILL_DIR = path.join(__dirname, '..');
const MAX_DEPTH = 3;

/**
 * Load a JSON config file
 */
function loadConfig(configName) {
  const configPath = path.join(SKILL_DIR, 'shared', 'config', `${configName}.json`);

  if (!fs.existsSync(configPath)) {
    throw new Error(`Config file not found: ${configPath}`);
  }

  const content = fs.readFileSync(configPath, 'utf8');
  return JSON.parse(content);
}

/**
 * Replace {{variables}} in content with values from config
 */
function replaceVariables(content, config) {
  let result = content;

  // Replace all {{variable}} patterns
  result = result.replace(/\{\{(\w+)\}\}/g, (match, varName) => {
    if (config.hasOwnProperty(varName)) {
      return config[varName];
    }
    // Keep the placeholder if variable not found in config
    console.warn(`Warning: Variable {{${varName}}} not found in config`);
    return match;
  });

  return result;
}

/**
 * Resolve a single include directive
 */
function resolveInclude(directive, config, depth, processedFiles) {
  const { type, filePath, configName } = directive;

  // Prevent infinite recursion
  if (depth > MAX_DEPTH) {
    throw new Error(`Max include depth (${MAX_DEPTH}) exceeded`);
  }

  // Resolve file path relative to skill directory
  const fullPath = path.join(SKILL_DIR, filePath);

  // Detect circular includes
  if (processedFiles.has(fullPath)) {
    throw new Error(`Circular include detected: ${filePath}`);
  }

  if (!fs.existsSync(fullPath)) {
    throw new Error(`Include file not found: ${fullPath}`);
  }

  // Read the included file
  let content = fs.readFileSync(fullPath, 'utf8');

  // Mark as processed
  processedFiles.add(fullPath);

  // If it's a parameterized include, load config and replace variables
  if (type === 'with-config') {
    if (!configName) {
      throw new Error(`Config name required for @include-with-config: ${filePath}`);
    }

    const domainConfig = config || loadConfig(configName);
    content = replaceVariables(content, domainConfig);
  }

  // Recursively resolve nested includes
  content = resolveIncludes(content, config, depth + 1, processedFiles);

  // Remove from processed to allow reuse in different branches
  processedFiles.delete(fullPath);

  return content;
}

/**
 * Parse an include directive
 */
function parseDirective(directiveText) {
  // Pattern: <!-- @include(-with-config)? path/to/file.md (config=name)? -->
  const match = directiveText.match(/<!--\s*@include(-with-config)?\s+(.+?)\s*(?:config=(\S+))?\s*-->/);

  if (!match) {
    return null;
  }

  return {
    full: match[0],
    type: match[1] ? 'with-config' : 'simple',
    filePath: match[2].trim(),
    configName: match[3] || null
  };
}

/**
 * Resolve all includes in content
 */
function resolveIncludes(content, config = null, depth = 0, processedFiles = new Set()) {
  // Find all include directives
  const includePattern = /<!--\s*@include(-with-config)?\s+.+?\s*(?:config=\S+)?\s*-->/g;

  let result = content;
  let match;

  while ((match = includePattern.exec(content)) !== null) {
    const directive = parseDirective(match[0]);

    if (!directive) {
      console.warn(`Warning: Could not parse directive: ${match[0]}`);
      continue;
    }

    try {
      const resolvedContent = resolveInclude(directive, config, depth, processedFiles);
      result = result.replace(directive.full, resolvedContent);
    } catch (error) {
      console.error(`Error resolving include: ${directive.filePath}`);
      console.error(error.message);
      // Keep the directive in place on error
    }
  }

  return result;
}

/**
 * Main function
 */
function main() {
  const args = process.argv.slice(2);

  if (args.length === 0) {
    console.error('Usage: node resolve-includes.js <template-file> [output-file]');
    console.error('');
    console.error('Example:');
    console.error('  node resolve-includes.js templates/TEMPLATE_BUSINESS_CONTINUITY.md expanded.md');
    process.exit(1);
  }

  const templatePath = args[0];
  const outputPath = args[1] || null;

  if (!fs.existsSync(templatePath)) {
    console.error(`Error: Template file not found: ${templatePath}`);
    process.exit(1);
  }

  try {
    // Read template
    const content = fs.readFileSync(templatePath, 'utf8');

    // Resolve includes
    const expanded = resolveIncludes(content);

    // Output
    if (outputPath) {
      fs.writeFileSync(outputPath, expanded, 'utf8');
      console.log(`Template expanded successfully: ${outputPath}`);

      // Show statistics
      const originalLines = content.split('\n').length;
      const expandedLines = expanded.split('\n').length;
      console.log(`Lines: ${originalLines} → ${expandedLines} (+${expandedLines - originalLines})`);
    } else {
      console.log(expanded);
    }
  } catch (error) {
    console.error(`Error: ${error.message}`);
    process.exit(1);
  }
}

// Run if called directly
if (require.main === module) {
  main();
}

module.exports = { resolveIncludes, loadConfig, replaceVariables };
